### ISSUE TITLE
REST Client's mapper does not work with generics and uses HashMap as fallback

### ISSUE THREAD

**Issue by progaddict** [OpenLiberty/guide-microprofile-rest-client#67]
### bug description
Consider the following remote service's interface:
```java
@Path("stock")
public interface MyClient {

    @GET
    @Path("{stockName}/chart/date/{date}")
    List<ChartEntry> getChart(
                                @PathParam("stockName") final String stockName,
                                @PathParam("date") final String date
    );
}
```
which returns a list of following POJOs:
```java
public class ChartEntry {
    private String date;
    private String minute;
    private Double close;
    // ... getters and setters ...
}
```
(see https://github.com/progaddict/guide-microprofile-rest-client/commit/80b2a0dca2ee2dcdd7bd0547ce6c3c251e38f179)

The MP REST client fails to map HTTP response to the generic type `List<ChartEntry>`.
It falls back to `List<java.util.HashMap>` which cases the following exception:
```
java.util.HashMap cannot be cast to io.openliberty.guides.inventory.client.ChartEntry
```

### steps to reproduce
* `git clone -b use-complex-type-in-rest-client git@github.com:progaddict/guide-microprofile-rest-client.git`
* `cd guide-microprofile-rest-client`
* `mvn liberty:stop-server && mvn clean && mvn -DskipTests install liberty:start-server`
* go to `http://localhost:9080/inventory/systems/localhost`
* observe the error message: `java.lang.ClassCastException: java.util.HashMap cannot be cast to io.openliberty.guides.inventory.client.ChartEntry`
* verify that the response is a valid [JSON](https://api.iextrading.com/1.0/stock/AAPL/chart/date/20190205)

### expected behaviour
* the list items are cast to `ChartEntry`
* they are logged into `console.log` (with [err] level)

**Comment by progaddict** [comment]
### hypothesis
the bug is related to the Java's type erasure mechanism.
the client is generated at runtime and therefore the `RestClientBuilder` sees a `List<Object>`.
it does not see the original `ChartEntry` class which has been erased by the compiler and therefore falls back to `HashMap`

**Comment by progaddict** [comment]
it's interesting though that the MP REST Client's team uses generic types in their [example](https://github.com/eclipse/microprofile-rest-client):
```java
 @Path("/movies")
 public interface MovieReviewService {
     @GET
     Set<Movie> getAllMovies();

     @GET
     @Path("/{movieId}/reviews")
     Set<Review> getAllReviews( @PathParam("movieId") String movieId );

     @GET
     @Path("/{movieId}/reviews/{reviewId}")
     Review getReview( @PathParam("movieId") String movieId, @PathParam("reviewId") String reviewId );

     @POST
     @Path("/{movieId}/reviews")
     String submitReview( @PathParam("movieId") String movieId, Review review );

     @PUT
     @Path("/{movieId}/reviews/{reviewId}")
     Review updateReview( @PathParam("movieId") String movieId, @PathParam("reviewId") String reviewId, Review review );
 }
```

you guys also have generics in your [example](https://openliberty.io/blog/2018/01/31/mpRestClient.html):
```java
@Path("/playlist")
@Consumes("application/json")
public interface MusicPlaylistService {

    @GET
    List<String> getPlaylistNames();

    @GET
    @Path("/{playlistName}")
    List<Song> getPlaylist(@PathParam("playlistName") String name)
        throws UnknownPlaylistException;

    @POST
    @Path("/{playlistName}")
    long newPlayList(@PathParam("playlistName") String name, List<Song> playlist)
        throws PlaylistAlreadyExistsException;

    @PUT
    @Path("/{playlistName}")
    long updatePlayList(@PathParam("playlistName") String name, List<Song> playlist)
        throws UnknownPlaylistException;
}
```

I am curious. How do guys you get the information about the underlying `List`'s type at runtime when it has already been erased by the `javac`? Am I missing something here?

**Comment by progaddict** [comment]
even if it is a fundamental problem caused by the erasure mechanism, there should still be a work around.
because many real world APIs produce a list of some sort of items.
so there should be a way to consume such APIs.

**Comment by progaddict** [comment]
OK, there is a work around! :bulb: 

Since Java arrays are not subject to type erasure, they can be used as a return type and then the MP REST Client's runtime will be able to determine the type of the item in the collection:
https://github.com/progaddict/guide-microprofile-rest-client/commit/327e6e618c14e1b0972fd321c7dcf6aadc2aa081

However, this solution is suboptimal because one needs to constantly convert returned arrays to a corresponding collection type (i.e. `List<T>` or `Set<T>`):
```java
final List<ChartEntry> chart = Arrays.asList(myClient.getChart("AAPL", "20190205"));
```
which is somewhat cumbersome :disappointed: 

therefore I am still interested to know if there is a better solution :question:

**Comment by andymc12** [comment]
Hi @progaddict - sorry for the slow reply.  This is most likely a bug in Liberty's JSON-B JAX-RS provider (see Open Liberty issue [6521](
https://github.com/OpenLiberty/open-liberty/issues/6521).  The provider should be using the parameterized type when converting from JSON to object, but instead is only using the class (i.e. List, Set, Map, etc.) - so JSON-B doesn't know what to convert the collection values to.

This should work when using EE7 technology (prior to JSON-B) because we were using a different mechanism for JSON-to/from-Object conversion.

I'll try to have a fix for this shortly.  Thanks for opening the issue!

**Comment by progaddict** [comment]
Hi @andymc12 

Thank you for your reply.

I have not fully understood how this is going to be fixed... The generation of the HTTP client happens in **runtime**, doesn't it? So the `RestClientBuilder` upon `.build()` call sees the following interface:
```java
@Path("stock")
public interface MyClient {

    @GET
    @Path("{stockName}/chart/date/{date}")
    List<Object> getChart(
            @PathParam("stockName") final String stockName,
            @PathParam("date") final String date
    );
}
```
The `ChartEntry` class has been erased by the `javac` and there is now just `Object`. So how does the actual type can be restored?

I'd suggest to continue our discussion in https://github.com/eclipse/microprofile-rest-client/issues/157 in order not to duplicate messages here and there.

**Comment by andymc12** [comment]
The bug fix for this is now in OpenLiberty's master branch and should be part of the next release.  This issue should remain open until the guide has been updated to include the latest OL release.

**Comment by fachat** [comment]
Thanks for fixing this. 
Is it now also made sure that such a regression is tested for in future releases? It just happened to me when I changed my docker base container from liberty javaee7 to javaee8 and I had a bad time looking for the cause...

**Comment by andymc12** [comment]
Hi @fachat - yes, we added some automated testing here:
https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.microprofile.rest.client_fat/test-applications/collectionsApp/src/mpRestClient10/collections/CollectionsTestServlet.java

This bug was specific to Java EE 8 because it was in our JSON-B provider code (which was introduced as part of JAX-RS 2.1 / Java EE 8 - in Java EE 7, we used a Jackson-based provider).  This should be fixed in Open Liberty 19.0.0.2.

Thanks!
